src/lib/tls/tls13/tls_extensions_key_share.cpp

```diff
@@ -426,22 +449,27 @@ bool Key_Share::empty() const {
    return std::visit([](const auto& key_share) { return key_share.empty(); }, m_impl->key_share);
 }
 
-secure_vector<uint8_t> Key_Share::exchange(const Key_Share& peer_keyshare,
-                                           const Policy& policy,
-                                           Callbacks& cb,
-                                           RandomNumberGenerator& rng) const {
-   return std::visit(
-      overloaded{[&](const Key_Share_ClientHello& ch, const Key_Share_ServerHello& sh) {
-                    return ch.exchange(sh, policy, cb, rng);
-                 },
-                 [&](const Key_Share_ServerHello& sh, const Key_Share_ClientHello& ch) {
-                    return sh.exchange(ch, policy, cb, rng);
-                 },
-                 [](const auto&, const auto&) -> secure_vector<uint8_t> {
-                    throw Invalid_Argument("can only exchange with ServerHello and ClientHello Key_Share");
-                 }},
-      m_impl->key_share,
-      peer_keyshare.m_impl->key_share);
+std::unique_ptr<Key_Share> Key_Share::create_as_encapsulation(Group_Params selected_group,
+                                                              const Key_Share& client_keyshare,
+                                                              const Policy& policy,
+                                                              Callbacks& cb,
+                                                              RandomNumberGenerator& rng) {
+   return std::unique_ptr<Key_Share>(new Key_Share(selected_group, client_keyshare, policy, cb, rng));
+}
+
+secure_vector<uint8_t> Key_Share::decapsulate(const Key_Share& server_keyshare,
+                                              const Policy& policy,
+                                              Callbacks& cb,
+                                              RandomNumberGenerator& rng) {
+   return std::visit(overloaded{[&](Key_Share_ClientHello& ch, const Key_Share_ServerHello& sh) {
```

Comment:
I really hope pattern matching makes it in C++26 ...

Reply:
Yeah! Though, as stated in the "future work" above: The internal implementation of `Key_Share` may be more complicated than it needs to.